### PR TITLE
test/cls/rgw: add index transaction simulator to model bucket stats

### PR DESCRIPTION
**Pull request by cbodley**
# ceph_test_cls_rgw_stats

a simulator that generates lots of create/delete operations against a set of object names, and tracks the expected bucket stats for comparison

each operation has a 70% chance to complete successfully, a 10% chance to cancel, 10% chance to suggest update, and 10% chance to suggest removal

TODO:
- [x] simulate multipart uploads (covers the `remove_objs` case of `cls_rgw_bucket_complete_op`)

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cbodley** [comment]
still need to figure out a good way to model multipart uploads. i don't *think* we need to model versioned objects, because only the plain entries are accounted for in stats

File: src/cls/rgw/cls_rgw_types.h [hunk 1]
```diff
@@ -736,6 +756,21 @@ struct rgw_bucket_category_stats {
...
 WRITE_CLASS_ENCODER(rgw_bucket_category_stats)
 
+inline bool operator==(const rgw_bucket_category_stats& lhs,
+                       const rgw_bucket_category_stats& rhs) {
+  return lhs.total_size == rhs.total_size
+      && lhs.total_size_rounded == rhs.total_size_rounded
+      && lhs.num_entries == rhs.num_entries
+      && lhs.actual_size == rhs.actual_size;
+}
+inline bool operator!=(const rgw_bucket_category_stats& lhs,
```
Comment (ivancich) [inline src/cls/rgw/cls_rgw_types.h:766]: You don't want to define one in terms of the other? ;)

**Review by ivancich — APPROVED** [review]
Very nice PR! Cleaning up some technical debt along the way!

**Comment by cbodley** [comment]
rebased and updated the rgw/verify suite to run it

**Comment by cbodley** [comment]
passed in https://pulpito.ceph.com/cbodley-2021-11-19_19:49:35-rgw-wip-cbodley-testing-distro-basic-smithi/

**Pull request merged**